Cython/Compiler/Nodes.py

```diff
@@ -1226,11 +1227,15 @@ class TemplatedTypeNode(CBaseTypeNode):
     name = None
 
     def _analyse_template_types(self, env, base_type):
-        require_python_types = base_type.python_type_constructor_name in (
-            'typing.Optional',
-            'dataclasses.ClassVar',
-        )
-        in_c_type_context = env.in_c_type_context and not require_python_types
+
+        required_types = {
+            'typing.Optional': lambda t: t.is_pyobject or t.is_buffer or t.is_memoryviewslice,
+            'dataclasses.ClassVar': lambda t: t.is_pyobject,
+        }
```

Comment:
This looks like we should extract the dict into a constant. Creating a dict and two function objects just to (maybe) evaluate a condition seems unnecessarily costly. Also, since we're using the same condition as for argument types further up, there should probably be a helper function `type_supports_Optional()`.

Or, maybe even better: why not add a method on type classes (`can_be_optional()`) which those types that support `Optional` can override to return True? Similar to `is_simple()`, `is_complete()`, etc.

Reply:
I did not forgot about this. I did not have enough time...

Reply:
fixed.